lib/libopenswan/atoasr.3
.TH "IPSEC_ATOASR" 3 "11 June 2001"
.\" RCSID $Id: atoasr.3,v 1.10 2004/04/09 18:00:38 mcr Exp $
.SH NAME
ipsec_atoasr, ipsec_rangetoa \- convert ASCII to Internet address, subnet, or range, convert Internet address range to ASCII
.SH SYNOPSIS
.B "#include <freeswan.h>
.sp
.B "const char *atoasr(const char *src, size_t srclen,"
.ti +1c
.B "char *type, struct in_addr *addrs);"
.br
.B "size_t rangetoa(struct in_addr *addrs, int format,
.ti +1c
.B "char *dst, size_t dstlen);"
.SH DESCRIPTION
These functions are obsolete;
there is no current equivalent,
because so far they have not proved useful.
.PP
.I Atoasr
converts an ASCII address, subnet, or address range
into a suitable combination of binary addresses
(in network byte order).
.I Rangetoa
converts an address range back into ASCII,
using dotted-decimal form for the addresses
(the other reverse conversions are handled by
.IR ipsec_addrtoa (3)
and
.IR ipsec_subnettoa (3)).
.PP
A single address can be any form acceptable to
.IR ipsec_atoaddr (3):
dotted decimal, DNS name, or hexadecimal number.
A subnet
specification uses the form \fInetwork\fB/\fImask\fR
interpreted by
.IR ipsec_atosubnet (3).
.PP
An address range is two
.IR ipsec_atoaddr (3)
addresses separated by a
.B ...
delimiter.
If there are four dots rather than three, the first is taken as
part of the begin address,
e.g. for a complete DNS name which ends with
.B .
to suppress completion attempts.
The begin address of a range must be
less than or equal to the end address.
.PP
The
.I srclen
parameter of
.I atoasr
specifies the length of the ASCII string pointed to by
.IR src ;
it is an error for there to be anything else
(e.g., a terminating NUL) within that length.
As a convenience for cases where an entire NUL-terminated string is
to be converted,
a
.I srclen
value of
.B 0
is taken to mean
.BR strlen(src) .
.PP
The
.I type
parameter of
.I atoasr
must point to a
.B char
variable used to record which form was found.
The
.I addrs
parameter must point to a two-element array of
.B "struct in_addr"
which receives the results.
The values stored into
.BR *type ,
and the corresponding values in the array, are:
.PP
.ta 3c +2c +3c
	*type	addrs[0]	addrs[1]
.sp 0.8
address	\&\fB'a'\fR	address	-
.br
subnet	\&\fB's'\fR	network	mask
.br
range	\&\fB'r'\fR	begin	end
.PP
The
.I dstlen
parameter of
.I rangetoa
specifies the size of the
.I dst
parameter;
under no circumstances are more than
.I dstlen
bytes written to
.IR dst .
A result which will not fit is truncated.
.I Dstlen
can be zero, in which case
.I dst
need not be valid and no result is written,
but the return value is unaffected;
in all other cases, the (possibly truncated) result is NUL-terminated.
The
.I freeswan.h
header file defines a constant,
.BR RANGETOA_BUF ,
which is the size of a buffer just large enough for worst-case results.
.PP
The
.I format
parameter of
.I rangetoa
specifies what format is to be used for the conversion.
The value
.B 0
(not the ASCII character
.BR '0' ,
but a zero value)
specifies a reasonable default,
and is in fact the only format currently available.
This parameter is a hedge against future needs.
.PP
.I Atoasr
returns NULL for success and
a pointer to a string-literal error message for failure;
see DIAGNOSTICS.
.I Rangetoa
returns
.B 0
for a failure, and otherwise
always returns the size of buffer which would 
be needed to
accommodate the full conversion result, including terminating NUL;
it is the caller's responsibility to check this against the size of
the provided buffer to determine whether truncation has occurred.
.SH SEE ALSO
ipsec_atoaddr(3), ipsec_atosubnet(3)
.SH DIAGNOSTICS
Fatal errors in
.I atoasr
are:
empty input;
error in
.IR ipsec_atoaddr (3)
or
.IR ipsec_atosubnet (3)
during conversion;
begin address of range exceeds end address.
.PP
Fatal errors in
.I rangetoa
are:
unknown format.
.SH HISTORY
Written for the FreeS/WAN project by Henry Spencer.
.SH BUGS
The restriction of error reports to literal strings
(so that callers don't need to worry about freeing them or copying them)
does limit the precision of error reporting.
.PP
The error-reporting convention lends itself
to slightly obscure code,
because many readers will not think of NULL as signifying success.
A good way to make it clearer is to write something like:
.PP
.RS
.nf
.B "const char *error;"
.sp
.B "error = atoasr( /* ... */ );"
.B "if (error != NULL) {"
.B "        /* something went wrong */"
.fi
.RE


CHANGES

Note that listings are in chronological order of release times, not in order
of version numbers, so you will find 2.5.x and 2.6.x releases intersperced.

See also docs/KNOWN_BUGS and http://bugs.openswan.org/

v2.6.32 (unreleased)
* Remove -DLEAK_DETECTIVE forced by default [Tuomo]
* Makefile.inc now uses USE_LEAK_DETECTIVE?=false [Paul]
* Add -DCOMPILER_HAS_NO_PRINTF_LIKE to support an arm-elf-gcc compiler [Paul]
* If pluto is started with --nofork, then also disable nhelpers [Paul]
* Added HAVE_NO_FORK?= option to Makefile.inc (default false) [Paul]
* Bugtracker bugs fixed:
   #1095 Local packets are dropped on ipsec device when marking packets in
         OUTPUT chain [Wolfgang Nothdurft]
   #1160 init.d script not reporting correct exit status on config parse
         error [James Mead]

v2.6.31 (October 18, 2010)
* Fix for OCSP compile of commit 934ce6c9443832c [Paul]

v2.6.30 (October 18, 2010)
* Fix for: ASSERTION FAILED at connections.c:1579: c->kind == CK_TEMPLATE [Paul]
* Add --listen/listen= option to limit listening to a single IP [Paul]
* sa_policy_bit_names was missing ModeConfig DNS and WINS bit names [Paul]
* SAREF: Add sareftrack=<no|yes|conntrack> connection option [Paul]
* Fix for "handling event EVENT_RETRANSMIT for <invalid>" [Paul]
* Fix for specifying protport=47 (GRE has no ports) [Paul]
* Don't fill traffic selector struct in IKEv2 child SA when not recv'd [Paul]
* Add geode-aes to the list of crypto modules to load [Paul]
* Don't install ipsec init script as /etc/init.d/setup [David]
* Don't create rc.? symlinks - let user use chkconfig etc instead [Paul]
* Bugtracker bugs fixed:
   # 252/619 more than 20 payloads in message; ignored [paul]
   # 690 ipsec lwdnsq --help not implemented
   # 860 Port --random for newhostkey [Paul]
   #1005 Incorrect message "R_U_THERE_ACK has unexpected sequence number" [Mike]
   #1040: Fix to compile without DEBUG [Paul]
   #1054: Startup warning: "ignored obsolete keyword (null)" [Michael Smith]
   #1112: Prototypes only, if function enabled in c-source with KLIPS or
          PFKEY [Henry N.]
   #1115: Fix various warnings u_char * vs. char * for sscanf,... [Henry N.]
   #1149: pluto uses empty NAT_OA as IDci, Server behind NAT and non natted
          Windows XP [Wolfgang Nothdurft]
   #1151: The ipsec module is not removed by 'ipsec setup stop' [Paul]

v2.6.29 (September 27, 2010)
! This release is made for CVE-2010-3302 and CVE-2010-3308
! Duplicate CVE's issues for these are CVE-2010-3752 and CVE-2010-3753
* XAUTH: Avoid buffer overflow in CISCO DNS info [dhr/paul]
         Avoid shell problems with single quotes CISCO DNS paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3302
* XAUTH: Avoid buffer overflow in CISCO BANNER [dhr/paul]
         Avoid shell problems with single quotes in CISCO paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3308
* NETKEY: Fix for spurious %hold netlink-acquires [Paul/dhr]
* KLIPS: Fix compiling on 2.6.18 based RHEL5 kernels [Paul]
* Various fixes based on automated source code review [dhr]
* SAREF: Updated for 2.6.35 kernel [Harald]
* KLIPS: Updated for 2.6.35 kernel [Harald]
* PACKAGING Use Epoch 1: for Debian/Ubuntu [Simon]
* MAST: fix iptables rule "leak" on rekey [Bart]
* MAST: use only the most recent iptables rule [Bart]
* pluto: restrict rekeymargin to be smaller than salifetime [Bart]
* MAST: ensure we don't end up with mtu=0 on mast0 [Bart/Paul]
* MAST: enforce outgoing tunnel policy [Bart]
* MAST: use addflow pfkey command to set policy on tunnel SAs [Bart]
* Added a new pfkey flag, POLICYONLY, to the ADDFLOW command [Bart]
* MAST: allow for setting of policy for inbound SAs [Bart]
* MAST: favour deleting an SA even if the pfkey op failed [Bart]
* HAVESTATSD: Log new phase2 messages as a result of a rekey [Paul]
* MAST: use iptables --comment to show the conn name [Bart]
* VNET: differentiate instantiation of road warriors and vnet [Paul]
* Log LEAK_DETECTIVE and HAVE_LIBNSS support on startup [Paul]
* [IKEv2] connections were broken since 2.6.25 [Avesh]
* MAST: new "ipsec policy" command replaces "ipsec eroute" [Bart]
* Fix SElinux warning in realsetup (bz628879) [Avesh]
* Support for SHA2_256 in IKEv2 (bz621790) [Avesh]
* IKEv2: Fix for using MD5 and PRF conversion function [Avesh]
* SAREF: Improved workaround for rp_filter [Bart]
* NSS: Increase minimum nss for rhbz#453577 [Paul]
  (this allows us to revert workaround in git 6c8ff2791d1)
* SAREF: Added /proc/net/ipsec/saref that shows kernel patch state [Bart]
* SAREF: Add SAref checks to ipsec verify [bart/paul]
* A bunch of Makefile.inc variables could not be overridden properly [Paul]
* Bugtracker bugs fixed:
   #1042 NF_INET_LOCAL_OUT vs NF_IP_LOCAL_OUT still broken [David]
   #1102: Ensure KLIPS still works on older kernels [Paul]
   #1113: crypt_dh.c: Fix "warning: multi-line comment" [Tuomo]

v2.6.28 (July 29, 2010)
* MAST: Fix SAref vmalloc() call that could cause a kernel panic [Bart]
* SAREF: rework exported interface to avoid might_sleep() during rcu lock [Bart]
* SAREF: Use the correct mask when updating nfmark in rcv path [Bart]
* MAST: Cleanup updown.mast iptables rule management [Bart]
* MAST: Rework mast init scripts to use conntrack [Bart]
* MAST: Remove iptables rules after SA is deleted [Bart]
* SAREF: Fix bug in stream-socket saref mode [Bart]
* SAREF: Ported to apply on Linux 2.6.34 [Harald]
* MAST: Fix for NAT-T mode (set ixs->outgoing_said in mast mode) [Bart]
* MAST: Fix module unload with mast [Bart]
* Allow rightsubnet=vnet:%priv rightprotoport=17/%any w/o right=%any [Paul]
* SAREF: Log SAref and SAbind capabilities on pluto startup [Paul]
* Log tunnel down with HAVE_STATSD as "down", not "unknown" [Paul]
* Changes to _realsetup.in for making the init script LSB compliant [Avesh]
  rhbz #594767
* _startnetkey update for DNS and NetworkManager [Avesh]
* SAREF: fix bug in stream-socket saref mode [Bart]

lib/libopenswan/atoasr.c
/*
 * convert from ASCII form of address/subnet/range to binary
 * Copyright (C) 1998, 1999  Henry Spencer.
 * 
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/lgpl.txt>.
 * 
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
 * License for more details.
 *
 * RCSID $Id: atoasr.c,v 1.10 2004/04/09 18:00:38 mcr Exp $
 */
#include "internal.h"
#include "openswan.h"

/*
 - atoasr - convert ASCII to address, subnet, or range
 */
const char *			/* NULL for success, else string literal */
atoasr(src, srclen, typep, addrsp)
const char *src;
size_t srclen;			/* 0 means "apply strlen" */
char *typep;			/* return type code:  'a', 's', 'r' */
struct in_addr addrsp[2];
{
	const char *punct;
	const char *stop;
	const char *oops;

	if (srclen == 0)
		srclen = strlen(src);
	if (srclen == 0)
		return "empty string";

	/* subnet is easy to spot */
	punct = memchr(src, '/', srclen);
	if (punct != NULL) {
		*typep = 's';
		return atosubnet(src, srclen, &addrsp[0], &addrsp[1]);
	}

	/* try for a range */
	stop = src + srclen;
	for (punct = src; (punct = memchr(punct, '.', stop - punct)) != NULL;
									punct++)
		if (stop - punct > 3 && *(punct+1) == '.' && *(punct+2) == '.')
			break;			/* NOTE BREAK OUT */
	if (punct == NULL) {
		/* didn't find the range delimiter, must be plain address */
		*typep = 'a';
		return atoaddr(src, srclen, &addrsp[0]);
	}

	/* looks like a range */
	*typep = 'r';
	if (stop - punct > 4 && *(punct+3) == '.')
		punct++;		/* first dot is trailing dot of name */
	oops = atoaddr(src, punct - src, &addrsp[0]);
	if (oops != NULL)
		return oops;
	oops = atoaddr(punct+3, stop - (punct+3), &addrsp[1]);
	if (oops != NULL)
		return oops;
	if (ntohl(addrsp[0].s_addr) > ntohl(addrsp[1].s_addr))
		return "invalid range, begin > end";
	return NULL;
}



#ifdef ATOASR_MAIN

#include <stdio.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

void regress(void);

int
main(int argc, char *argv[])
{
	struct in_addr a[2];
	char buf[100];
	const char *oops;
	size_t n;
	char type;

	if (argc < 2) {
		fprintf(stderr, "Usage: %s {addr|net/mask|begin...end|-r}\n",
								argv[0]);
		exit(2);
	}

	if (strcmp(argv[1], "-r") == 0) {
		regress();
		fprintf(stderr, "regress() returned?!?\n");
		exit(1);
	}

	oops = atoasr(argv[1], 0, &type, a);
	if (oops != NULL) {
		fprintf(stderr, "%s: conversion failed: %s\n", argv[0], oops);
		exit(1);
	}
	switch (type) {
	case 'a':
		n = addrtoa(a[0], 0, buf, sizeof(buf));
		break;
	case 's':
		n = subnettoa(a[0], a[1], 0, buf, sizeof(buf));
		break;
	case 'r':
		n = rangetoa(a, 0, buf, sizeof(buf));
		break;
	default:
		fprintf(stderr, "%s: unknown type '%c'\n", argv[0], type);
		exit(1);
		break;
	}
	if (n > sizeof(buf)) {
		fprintf(stderr, "%s: reverse conversion of ", argv[0]);
		fprintf(stderr, "%s ", inet_ntoa(a[0]));
		fprintf(stderr, "%s", inet_ntoa(a[1]));
		fprintf(stderr, " failed: need %ld bytes, have only %ld\n",
						(long)n, (long)sizeof(buf));
		exit(1);
	}
	printf("%s\n", buf);

	exit(0);
}

struct rtab {
	char *input;
	char *output;			/* NULL means error expected */
} rtab[] = {
	{"1.2.3.0",			"1.2.3.0"},
	{"1.2.3.0/255.255.255.0",	"1.2.3.0/24"},
	{"1.2.3.0...1.2.3.5",		"1.2.3.0...1.2.3.5"},
	{"1.2.3.4.5",			NULL},
	{"1.2.3.4/",			NULL},
	{"1.2.3.4...",			NULL},
	{"1.2.3.4....",			NULL},
	{"localhost/32",		        "127.0.0.1/32"},
	{"localhost...127.0.0.3",	"127.0.0.1...127.0.0.3"},
	{"127.0.0.0...localhost",	"127.0.0.0...127.0.0.1"},
	{"127.0.0.3...localhost",	NULL},
	{NULL,				NULL}
};

void
regress(void)
{
	struct rtab *r;
	int status = 0;
	struct in_addr a[2];
	char in[100];
	char buf[100];
	const char *oops;
	size_t n;
	char type;

	for (r = rtab; r->input != NULL; r++) {
		strcpy(in, r->input);
		oops = atoasr(in, 0, &type, a);
		if (oops != NULL && r->output == NULL)
			{}		/* okay, error expected */
		else if (oops != NULL) {
			printf("`%s' atoasr failed: %s\n", r->input, oops);
			status = 1;
		} else if (r->output == NULL) {
			printf("`%s' atoasr succeeded unexpectedly '%c'\n",
							r->input, type);
			status = 1;
		} else {
			switch (type) {
			case 'a':
				n = addrtoa(a[0], 0, buf, sizeof(buf));
				break;
			case 's':
				n = subnettoa(a[0], a[1], 0, buf, sizeof(buf));
				break;
			case 'r':
				n = rangetoa(a, 0, buf, sizeof(buf));
				break;
			default:
				fprintf(stderr, "`%s' unknown type '%c'\n",
							r->input, type);
				n = 0;
				status = 1;
				break;
			}
			if (n > sizeof(buf)) {
				printf("`%s' '%c' reverse failed:  need %ld\n",
						r->input, type, (long)n);
				status = 1;
			} else if (n > 0 && strcmp(r->output, buf) != 0) {
				printf("`%s' '%c' gave `%s', expected `%s'\n",
					r->input, type, buf, r->output);
				status = 1;
			}
		}
	}
	exit(status);
}

#endif /* ATOASR_MAIN */
